packages/react-reconciler/src/ReactFiberBeginWork.js

```diff
@@ -1025,24 +1177,53 @@ function beginWork(
   workInProgress.expirationTime = NoWork;
 
   switch (workInProgress.tag) {
-    case IndeterminateComponent:
+    case IndeterminateComponent: {
+      const Component = workInProgress.type;
       return mountIndeterminateComponent(
         current,
         workInProgress,
+        Component,
         renderExpirationTime,
       );
-    case FunctionalComponent:
+    }
+    case FunctionalComponent: {
+      const Component = workInProgress.type;
       return updateFunctionalComponent(
         current,
         workInProgress,
+        Component,
         renderExpirationTime,
       );
-    case ClassComponent:
+    }
+    case FunctionalComponentLazy: {
+      const thenable = workInProgress.type;
+      const Component = (thenable._reactResult: any);
+      return updateFunctionalComponentLazy(
+        current,
+        workInProgress,
+        Component,
+        renderExpirationTime,
+      );
+    }
+    case ClassComponent: {
+      const Component = workInProgress.type;
```

Comment:
You already pass the resolved type as an argument. I think it would be better to do the same thing with props.

```js
const props = workInProgress.pendingProps;
return updateClassComponent(
        current,
        workInProgress,
        Component,
        props,
        renderExpirationTime,
      );
```

Then in the lazy form you do 
```js
const props = resolveDefaultProps(Component, workInProgress.pendingProps);
```

That way you don't need any special casing for null and separate `updateClassComponentLazy` helper and such.

Reply:
I think I'll still need `updateClassComponentLazy` to do the `defaultProps` resolution

Reply:
No, you just do it in the switch statement like how you resolve the constructor.

Reply:
Oh nvm I read it wrong

Reply:
The tricky thing is that I think I need to reset the memoized props back to the unresolved value so we can quickly bail out the next time.